README.md
# terraform-plugin-framework-type-time

[![PkgGoDev](https://pkg.go.dev/badge/github.com/bflad/terraform-plugin-framework-type-time)](https://pkg.go.dev/github.com/bflad/terraform-plugin-framework-type-time)

[RFC 3339](https://tools.ietf.org/html/rfc3339) timestamp string type and value implementations for [terraform-plugin-framework](https://github.com/hashicorp/terraform-plugin-framework). These types automatically include syntax validation and support Go standard library [`time.Time`](https://pkg.go.dev/time#Time) conversion.

## Status

_Experimental_: This project is not officially maintained by HashiCorp.

## Prerequisites

This Go module tracks recent versions of `terraform-plugin-framework` for Go version and interface compatibility.

## Getting Started

### Schema

Replace usage of `types.StringType` in schema definitions with `timetypes.RFC3339Type{}`.

Given the previous schema attribute:

```go
tfsdk.Attribute{
    Required: true,
    Type:     types.StringType,
    // Potentially previous Validators
}
```

The updated schema attribute:

```go
tfsdk.Attribute{
    Required: true,
    Type:     timetypes.RFC3339Type{},
}
```

### Schema Data Model

Replace usage of `string`, `*string`, or `types.String` in schema data models with `timetypes.RFC3339`.

Given the previous schema data model:

```go
type ThingResourceModel struct {
    // ...
    Example types.String `tfsdk:"example"`
}
```

The updated schema data model:

```go
type ThingResourceModel struct {
    // ...
    Example timetypes.RFC3339 `tfsdk:"example"`
}
```

### Accessing Values

Similar to other value types, use the `IsNull()` and `IsUnknown()` methods to check whether the value is null or unknown. Use the `Time()` method to extract a known `time.Time` value.

### Writing Values

Create a `timetypes.RFC3339` by calling one of these functions:

- `RFC3339Null() RFC3339`: creates a `null` value.
- `RFC3339String(string, path.Path) (Value, diag.Diagnostics)`: creates a known value using the given `string` or returns validation errors if `string` is not in the expected RFC 3339 format.
- `RFC3339Time(time.Time) Value` creates a known value using the given `time.Time`.
- `RFC3339Unknown() Value`: creates an unknown value.

### Adding the Dependency

All functionality is located in the `github.com/bflad/terraform-plugin-framework-type-time/timetypes` package. Add this to relevant Go file `import` statements.

Run these Go module commands to fetch the latest version and ensure all module files are up to date.

```shell
go get github.com/bflad/terraform-plugin-framework-type-time@latest
go mod tidy
```


CHANGELOG.md
# 0.2.1 (October 3, 2022)

BUG FIXES:

* Renamed `RFC3339Value` to `RFC3339` to match previous changelog

# 0.2.0 (October 3, 2022)

BREAKING CHANGES:

* The `rfc3339type` package was moved to `timetypes`, the `Type` type to `RFC3339Type`, and the `Value` type to `RFC3339`.

NOTES:

* This Go module was moved to `github.com/bflad/terraform-plugin-framework-type-time`.

# 0.1.0 (September 22, 2022)

FEATURES:

* rfc3339type: Initial `Type` and `Value` types


rfc3339type/type_test.go
package rfc3339type_test

import (
	"context"
	"fmt"
	"strings"
	"testing"
	"time"

	"github.com/bflad/terraform-plugin-framework-type-rfc3339/rfc3339type"
	"github.com/google/go-cmp/cmp"
	"github.com/hashicorp/terraform-plugin-framework/attr"
	"github.com/hashicorp/terraform-plugin-framework/diag"
	"github.com/hashicorp/terraform-plugin-framework/path"
	"github.com/hashicorp/terraform-plugin-framework/types"
	"github.com/hashicorp/terraform-plugin-go/tftypes"
)

func TestTypeApplyTerraform5AttributePathStep(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		typ           rfc3339type.Type
		step          tftypes.AttributePathStep
		expected      any
		expectedError error
	}{
		"AttributeName": {
			typ:           rfc3339type.Type{},
			step:          tftypes.AttributeName("test"),
			expectedError: fmt.Errorf("cannot apply AttributePathStep tftypes.AttributeName to rfc3339type.Type"),
		},
		"ElementKeyInt": {
			typ:           rfc3339type.Type{},
			step:          tftypes.ElementKeyInt(1),
			expectedError: fmt.Errorf("cannot apply AttributePathStep tftypes.ElementKeyInt to rfc3339type.Type"),
		},
		"ElementKeyString": {
			typ:           rfc3339type.Type{},
			step:          tftypes.ElementKeyString("test"),
			expectedError: fmt.Errorf("cannot apply AttributePathStep tftypes.ElementKeyString to rfc3339type.Type"),
		},
		"ElementKeyValue": {
			typ:           rfc3339type.Type{},
			step:          tftypes.ElementKeyValue{},
			expectedError: fmt.Errorf("cannot apply AttributePathStep tftypes.ElementKeyValue to rfc3339type.Type"),
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got, err := testCase.typ.ApplyTerraform5AttributePathStep(testCase.step)

			if err != nil {
				if testCase.expectedError == nil {
					t.Fatalf("expected no error, got: %s", err)
				}

				if !strings.Contains(err.Error(), testCase.expectedError.Error()) {
					t.Fatalf("expected error %q, got: %s", testCase.expectedError, err)
				}
			}

			if err == nil && testCase.expectedError != nil {
				t.Fatalf("got no error, tfType: %s", testCase.expectedError)
			}

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestTypeEqual(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		typ      rfc3339type.Type
		other    attr.Type
		expected bool
	}{
		"nil": {
			typ:      rfc3339type.Type{},
			other:    nil,
			expected: false,
		},
		"rfc3339type.Type": {
			typ:      rfc3339type.Type{},
			other:    rfc3339type.Type{},
			expected: true,
		},
		"types.StringType": {
			typ:      rfc3339type.Type{},
			other:    types.StringType,
			expected: false,
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.typ.Equal(testCase.other)

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestTypeString(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		typ      rfc3339type.Type
		expected string
	}{
		"any": {
			typ:      rfc3339type.Type{},
			expected: "rfc3339type.Type",
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.typ.String()

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestTypeTerraformType(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		typ      rfc3339type.Type
		expected tftypes.Type
	}{
		"any": {
			typ:      rfc3339type.Type{},
			expected: tftypes.String,
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.typ.TerraformType(context.Background())

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestTypeValidate(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		typ            rfc3339type.Type
		terraformValue tftypes.Value
		schemaPath     path.Path
		expectedDiags  diag.Diagnostics
	}{
		"not-string": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.Bool, true),
			schemaPath:     path.Root("test"),
			expectedDiags: diag.Diagnostics{
				diag.NewAttributeErrorDiagnostic(
					path.Root("test"),
					"Invalid RFC 3339 Terraform Value",
					"An unexpected error occurred while attempting to read a RFC 3339 string from the Terraform value. "+
						"Please contact the provider developers with the following:\n\n"+
						"Error: can't unmarshal tftypes.Bool into *string, expected string",
				),
			},
		},
		"string-null": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, nil),
			schemaPath:     path.Root("test"),
		},
		"string-unknown": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, tftypes.UnknownValue),
			schemaPath:     path.Root("test"),
		},
		"string-value-invalid": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, "not-rfc3339-format"),
			schemaPath:     path.Root("test"),
			expectedDiags: diag.Diagnostics{
				diag.NewAttributeErrorDiagnostic(
					path.Root("test"),
					"Invalid RFC 3339 String Value",
					"An unexpected error occurred while converting a string value that was expected to be RFC 3339 format. "+
						"The RFC 3339 string format is YYYY-MM-DDTHH:MM:SSZ, such as 2006-01-02T15:04:05Z or 2006-01-02T15:04:05+07:00.\n\n"+
						"Error: parsing time \"not-rfc3339-format\" as \"2006-01-02T15:04:05Z07:00\": cannot parse \"not-rfc3339-format\" as \"2006\"",
				),
			},
		},
		"string-value-valid-offset-negative": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05-07:00"),
			schemaPath:     path.Root("test"),
		},
		"string-value-valid-offset-positive": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05+07:00"),
			schemaPath:     path.Root("test"),
		},
		"string-value-valid-z": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05Z"),
			schemaPath:     path.Root("test"),
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			diags := testCase.typ.Validate(context.Background(), testCase.terraformValue, testCase.schemaPath)

			if diff := cmp.Diff(diags, testCase.expectedDiags); diff != "" {
				t.Errorf("unexpected diagnostics difference: %s", diff)
			}
		})
	}
}

func TestTypeValueFromTerraform(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		typ            rfc3339type.Type
		terraformValue tftypes.Value
		expected       attr.Value
		expectedError  error
	}{
		"not-string": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.Bool, true),
			expected:       rfc3339type.UnknownValue(),
			expectedError:  fmt.Errorf("can't unmarshal tftypes.Bool into *string, expected string"),
		},
		"string-null": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, nil),
			expected:       rfc3339type.NullValue(),
		},
		"string-unknown": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, tftypes.UnknownValue),
			expected:       rfc3339type.UnknownValue(),
		},
		"string-value-invalid": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, "not-rfc3339-format"),
			expected:       rfc3339type.UnknownValue(),
			expectedError:  fmt.Errorf("parsing time \"not-rfc3339-format\" as \"2006-01-02T15:04:05Z07:00\": cannot parse \"not-rfc3339-format\" as \"2006\""),
		},
		"string-value-valid-offset-negative": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05-07:00"),
			expected:       rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", -7*60*60))),
		},
		"string-value-valid-offset-positive": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05+07:00"),
			expected:       rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", 7*60*60))),
		},
		"string-value-valid-z": {
			typ:            rfc3339type.Type{},
			terraformValue: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05Z"),
			expected:       rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got, err := testCase.typ.ValueFromTerraform(context.Background(), testCase.terraformValue)

			if err != nil {
				if testCase.expectedError == nil {
					t.Fatalf("expected no error, got: %s", err)
				}

				if !strings.Contains(err.Error(), testCase.expectedError.Error()) {
					t.Fatalf("expected error %q, got: %s", testCase.expectedError, err)
				}
			}

			if err == nil && testCase.expectedError != nil {
				t.Fatalf("got no error, tfType: %s", testCase.expectedError)
			}

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}


rfc3339type/value_test.go
package rfc3339type_test

import (
	"context"
	"testing"
	"time"

	"github.com/bflad/terraform-plugin-framework-type-rfc3339/rfc3339type"
	"github.com/google/go-cmp/cmp"
	"github.com/hashicorp/terraform-plugin-framework/attr"
	"github.com/hashicorp/terraform-plugin-framework/types"
	"github.com/hashicorp/terraform-plugin-go/tftypes"
)

func TestValueEqual(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		value    rfc3339type.Value
		other    attr.Value
		expected bool
	}{
		"nil": {
			value:    rfc3339type.NullValue(),
			other:    nil,
			expected: false,
		},
		"not-rfc3339type.Value": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			other:    types.String{Value: "2006-01-02T15:04:05Z"},
			expected: false,
		},
		"null-null": {
			value:    rfc3339type.NullValue(),
			other:    rfc3339type.NullValue(),
			expected: true,
		},
		"null-unknown": {
			value:    rfc3339type.NullValue(),
			other:    rfc3339type.UnknownValue(),
			expected: false,
		},
		"null-value": {
			value:    rfc3339type.NullValue(),
			other:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			expected: false,
		},
		"unknown-null": {
			value:    rfc3339type.UnknownValue(),
			other:    rfc3339type.NullValue(),
			expected: false,
		},
		"unknown-unknown": {
			value:    rfc3339type.UnknownValue(),
			other:    rfc3339type.UnknownValue(),
			expected: true,
		},
		"unknown-value": {
			value:    rfc3339type.UnknownValue(),
			other:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			expected: false,
		},
		"value-null": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			other:    rfc3339type.NullValue(),
			expected: false,
		},
		"value-unknown": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			other:    rfc3339type.UnknownValue(),
			expected: false,
		},
		"value-value-different": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			other:    rfc3339type.TimeValue(time.Date(2007, 2, 3, 16, 5, 6, 1, time.UTC)),
			expected: false,
		},
		"value-value-equal": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			other:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			expected: true,
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.value.Equal(testCase.other)

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestValueIsNull(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		value    rfc3339type.Value
		expected bool
	}{
		"null": {
			value:    rfc3339type.NullValue(),
			expected: true,
		},
		"unknown": {
			value:    rfc3339type.UnknownValue(),
			expected: false,
		},
		"value": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			expected: false,
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.value.IsNull()

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestValueIsUnknown(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		value    rfc3339type.Value
		expected bool
	}{
		"null": {
			value:    rfc3339type.NullValue(),
			expected: false,
		},
		"unknown": {
			value:    rfc3339type.UnknownValue(),
			expected: true,
		},
		"value": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			expected: false,
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.value.IsUnknown()

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestValueString(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		value    rfc3339type.Value
		expected string
	}{
		"null": {
			value:    rfc3339type.NullValue(),
			expected: "<null>",
		},
		"unknown": {
			value:    rfc3339type.UnknownValue(),
			expected: "<unknown>",
		},
		"value-offset-negative": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", -7*60*60))),
			expected: "\"2006-01-02T15:04:05-07:00\"",
		},
		"value-offset-positive": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", 7*60*60))),
			expected: "\"2006-01-02T15:04:05+07:00\"",
		},
		"value-z": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			expected: "\"2006-01-02T15:04:05Z\"",
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.value.String()

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestValueTime(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		value    rfc3339type.Value
		expected time.Time
	}{
		"null": {
			value:    rfc3339type.NullValue(),
			expected: time.Date(1, 1, 1, 0, 0, 0, 0, time.UTC),
		},
		"unknown": {
			value:    rfc3339type.UnknownValue(),
			expected: time.Date(1, 1, 1, 0, 0, 0, 0, time.UTC),
		},
		"value-offset-negative": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", -7*60*60))),
			expected: time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", -7*60*60)),
		},
		"value-offset-positive": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", 7*60*60))),
			expected: time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", 7*60*60)),
		},
		"value-z": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			expected: time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC),
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.value.Time()

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestValueToTerraformValue(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		value    rfc3339type.Value
		expected tftypes.Value
	}{
		"null": {
			value:    rfc3339type.NullValue(),
			expected: tftypes.NewValue(tftypes.String, nil),
		},
		"unknown": {
			value:    rfc3339type.UnknownValue(),
			expected: tftypes.NewValue(tftypes.String, tftypes.UnknownValue),
		},
		"value-offset-negative": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", -7*60*60))),
			expected: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05-07:00"),
		},
		"value-offset-positive": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.FixedZone("", 7*60*60))),
			expected: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05+07:00"),
		},
		"value-z": {
			value:    rfc3339type.TimeValue(time.Date(2006, 1, 2, 15, 4, 5, 0, time.UTC)),
			expected: tftypes.NewValue(tftypes.String, "2006-01-02T15:04:05Z"),
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.value.Type(context.Background())

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}

func TestValueType(t *testing.T) {
	t.Parallel()

	testCases := map[string]struct {
		value    rfc3339type.Value
		expected attr.Type
	}{
		"any": {
			value:    rfc3339type.NullValue(),
			expected: rfc3339type.Type{},
		},
	}

	for name, testCase := range testCases {
		name, testCase := name, testCase

		t.Run(name, func(t *testing.T) {
			t.Parallel()

			got := testCase.value.Type(context.Background())

			if diff := cmp.Diff(got, testCase.expected); diff != "" {
				t.Errorf("unexpected difference: %s", diff)
			}
		})
	}
}


rfc3339type/type.go
package rfc3339type

import (
	"context"
	"fmt"
	"time"

	"github.com/hashicorp/terraform-plugin-framework/attr"
	"github.com/hashicorp/terraform-plugin-framework/attr/xattr"
	"github.com/hashicorp/terraform-plugin-framework/diag"
	"github.com/hashicorp/terraform-plugin-framework/path"
	"github.com/hashicorp/terraform-plugin-go/tftypes"
)

// Ensure implementation satisfies expected interfaces.
var (
	_ tftypes.AttributePathStepper = Type{}
	_ attr.Type                    = Type{}
	_ xattr.TypeWithValidate       = Type{}
)

// Type implements the attr.Type interface for usage in schema definitions
// and data models.
type Type struct{}

// ApplyTerraform5AttributePathStep always returns an error as this type
// cannot be walked any further.
func (t Type) ApplyTerraform5AttributePathStep(step tftypes.AttributePathStep) (any, error) {
	return nil, fmt.Errorf("cannot apply AttributePathStep %T to %s", step, t.String())
}

// Equal returns true if the given type is Type.
func (t Type) Equal(o attr.Type) bool {
	_, ok := o.(Type)

	return ok
}

// String returns a human readable string of the type.
func (t Type) String() string {
	return "rfc3339type.Type"
}

// TerraformType always returns tftypes.String.
func (t Type) TerraformType(_ context.Context) tftypes.Type {
	return tftypes.String
}

// Validate ensures the value is always RFC 3339 conformant.
func (t Type) Validate(_ context.Context, terraformValue tftypes.Value, schemaPath path.Path) diag.Diagnostics {
	if terraformValue.IsNull() || !terraformValue.IsKnown() {
		return nil
	}

	var str string

	err := terraformValue.As(&str)

	if err != nil {
		return diag.Diagnostics{
			diag.NewAttributeErrorDiagnostic(
				schemaPath,
				"Invalid RFC 3339 Terraform Value",
				"An unexpected error occurred while attempting to read a RFC 3339 string from the Terraform value. "+
					"Please contact the provider developers with the following:\n\n"+
					"Error: "+err.Error(),
			),
		}
	}

	_, diags := StringValue(str, schemaPath)

	return diags
}

// ValueFromTerraform converts the tftypes.Value into a value.
func (t Type) ValueFromTerraform(_ context.Context, terraformValue tftypes.Value) (attr.Value, error) {
	if terraformValue.IsNull() {
		return NullValue(), nil
	}

	if !terraformValue.IsKnown() {
		return UnknownValue(), nil
	}

	var str string

	err := terraformValue.As(&str)

	if err != nil {
		return UnknownValue(), err
	}

	strTime, err := time.Parse(time.RFC3339, str)

	if err != nil {
		return UnknownValue(), err
	}

	return TimeValue(strTime), nil
}


rfc3339type/value.go
package rfc3339type

import (
	"context"
	"time"

	"github.com/hashicorp/terraform-plugin-framework/attr"
	"github.com/hashicorp/terraform-plugin-framework/diag"
	"github.com/hashicorp/terraform-plugin-framework/path"
	"github.com/hashicorp/terraform-plugin-go/tftypes"
)

// Ensure implementation satisfies expected interfaces.
var (
	_ attr.Value = Value{}
)

// NullValue returns a null Value.
func NullValue() Value {
	return Value{
		null: true,
	}
}

// StringValue returns a known Value or any errors while attempting
// to parse the string as RFC 3339 format.
func StringValue(s string, schemaPath path.Path) (Value, diag.Diagnostics) {
	t, err := time.Parse(time.RFC3339, s)

	if err != nil {
		return Value{
				unknown: true,
			}, diag.Diagnostics{
				diag.NewAttributeErrorDiagnostic(
					schemaPath,
					"Invalid RFC 3339 String Value",
					"An unexpected error occurred while converting a string value that was expected to be RFC 3339 format. "+
						"The RFC 3339 string format is YYYY-MM-DDTHH:MM:SSZ, such as 2006-01-02T15:04:05Z or 2006-01-02T15:04:05+07:00.\n\n"+
						"Error: "+err.Error(),
				),
			}
	}

	return Value{
		value: t,
	}, nil
}

// TimeValue returns a known Value with the given time.
func TimeValue(t time.Time) Value {
	return Value{
		value: t,
	}
}

// UnknownValue returns an unknown Value.
func UnknownValue() Value {
	return Value{
		unknown: true,
	}
}

// Value implements the attr.Value interface for usage in logic.
type Value struct {
	null    bool
	unknown bool
	value   time.Time
}

// Equal returns true if the given attr.Value matches the following:
//   - Is a Value type
//   - Has the same null, unknown, and value data
func (v Value) Equal(o attr.Value) bool {
	otherValue, ok := o.(Value)

	if !ok {
		return false
	}

	if otherValue.null != v.null {
		return false
	}

	if otherValue.unknown != v.unknown {
		return false
	}

	return otherValue.value.Equal(v.value)
}

// IsNull returns true if the Value represents a null Value.
func (v Value) IsNull() bool {
	return v.null
}

// IsUnknown returns true if the Value represents an unknown Value.
func (v Value) IsUnknown() bool {
	return v.unknown
}

// String returns a human readable string of the Value.
func (v Value) String() string {
	if v.null {
		return attr.NullValueString
	}

	if v.unknown {
		return attr.UnknownValueString
	}

	return `"` + v.value.Format(time.RFC3339) + `"`
}

// Time returns the time.Time of a Value.
func (v Value) Time() time.Time {
	return v.value
}

// ToTerraformValue converts the Value to a tftypes.String.
func (v Value) ToTerraformValue(_ context.Context) (tftypes.Value, error) {
	if v.null {
		return tftypes.NewValue(tftypes.String, nil), nil
	}

	if v.unknown {
		return tftypes.NewValue(tftypes.String, tftypes.UnknownValue), nil
	}

	return tftypes.NewValue(tftypes.String, v.value.Format(time.RFC3339)), nil
}

// Type returns Type.
func (v Value) Type(_ context.Context) attr.Type {
	return Type{}
}


timetypes/rfc3339_value.go
package timetypes

import (
	"context"
	"time"

	"github.com/hashicorp/terraform-plugin-framework/attr"
	"github.com/hashicorp/terraform-plugin-framework/diag"
	"github.com/hashicorp/terraform-plugin-framework/path"
	"github.com/hashicorp/terraform-plugin-go/tftypes"
)

// Ensure implementation satisfies expected interfaces.
var (
	_ attr.Value = RFC3339Value{}
)

// RFC3339Null returns a null RFC3339Value.
func RFC3339Null() RFC3339Value {
	return RFC3339Value{
		null: true,
	}
}

// RFC3339String returns a known RFC3339Value or any errors while attempting
// to parse the string as RFC 3339 format.
func RFC3339String(s string, schemaPath path.Path) (RFC3339Value, diag.Diagnostics) {
	t, err := time.Parse(time.RFC3339, s)

	if err != nil {
		return RFC3339Value{
				unknown: true,
			}, diag.Diagnostics{
				diag.NewAttributeErrorDiagnostic(
					schemaPath,
					"Invalid RFC 3339 String Value",
					"An unexpected error occurred while converting a string value that was expected to be RFC 3339 format. "+
						"The RFC 3339 string format is YYYY-MM-DDTHH:MM:SSZ, such as 2006-01-02T15:04:05Z or 2006-01-02T15:04:05+07:00.\n\n"+
						"Error: "+err.Error(),
				),
			}
	}

	return RFC3339Value{
		value: t,
	}, nil
}

// RFC3339Time returns a known RFC3339Value with the given time.
func RFC3339Time(t time.Time) RFC3339Value {
	return RFC3339Value{
		value: t,
	}
}

// RFC3339Unknown returns an unknown RFC3339Value.
func RFC3339Unknown() RFC3339Value {
	return RFC3339Value{
		unknown: true,
	}
}

// RFC3339Value implements the attr.Value interface for usage in logic.
type RFC3339Value struct {
	null    bool
	unknown bool
	value   time.Time
}

// Equal returns true if the given attr.Value matches the following:
//   - Is a RFC3339Value type
//   - Has the same null, unknown, and value data
func (v RFC3339Value) Equal(o attr.Value) bool {
	otherValue, ok := o.(RFC3339Value)

	if !ok {
		return false
	}

	if otherValue.null != v.null {
		return false
	}

	if otherValue.unknown != v.unknown {
		return false
	}

	return otherValue.value.Equal(v.value)
}

// IsNull returns true if the RFC3339Value represents a null Value.
func (v RFC3339Value) IsNull() bool {
	return v.null
}

// IsUnknown returns true if the RFC3339Value represents an unknown Value.
func (v RFC3339Value) IsUnknown() bool {
	return v.unknown
}

// String returns a human readable string of the RFC3339Value.
func (v RFC3339Value) String() string {
	if v.null {
		return attr.NullValueString
	}

	if v.unknown {
		return attr.UnknownValueString
	}

	return `"` + v.value.Format(time.RFC3339) + `"`
}

// Time returns the time.Time of a RFC3339Value.
func (v RFC3339Value) Time() time.Time {
	return v.value
}

// ToTerraformValue converts the RFC3339Value to a tftypes.String.
func (v RFC3339Value) ToTerraformValue(_ context.Context) (tftypes.Value, error) {
	if v.null {
		return tftypes.NewValue(tftypes.String, nil), nil
	}

	if v.unknown {
		return tftypes.NewValue(tftypes.String, tftypes.UnknownValue), nil
	}

	return tftypes.NewValue(tftypes.String, v.value.Format(time.RFC3339)), nil
}

// Type returns the attr.Type of RFC3339Value.
func (v RFC3339Value) Type(_ context.Context) attr.Type {
	return RFC3339Type{}
}
